source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,1259 @@
+
+    The node's quorum will consist of nodes whos sum of stakes will
+    be at least min_amount, or less if MAX_NODES_IN_QUORUM has been reached.
+
+    Params:
+        node_key = the key of the node for which to generate the quorum
+        node_stake = the stake of the node for which to generate the quorum
+        stakes = the list of stakes in descending order
+        min_amount = the minimum amount a node's quorum's sum of stake should
+                     be reached (unless MAX_NODES_IN_QUORUM is reached first)
+        rand_seed = the source of randomness
+
+    Notes:
+        dice() should be replaced / improved to be more efficient,
+        see also https://issues.dlang.org/show_bug.cgi?id=5849
+
+*******************************************************************************/
+
+private QuorumConfig buildQuorumImpl (PublicKey node_key, Amount node_stake,
+    in NodeStake[] stakes, const Amount min_amount, in Hash rand_seed)
+{
+    QuorumConfig quorum;
+    Amount quorum_sum;  // sum of the staked amount of the quorum for this node
+
+    // to filter out duplicates generated by dice()
+    auto added_nodes = BitField!uint(stakes.length);
+    auto rnd_gen = getGenerator(node_key, rand_seed);
+
+    // node must have itself in the quorum set
+    quorum.nodes ~= node_key;
+    if (!quorum_sum.add(node_stake))
+        assert(0);
+
+    while (quorum_sum < min_amount && quorum.nodes.length < MAX_NODES_IN_QUORUM)
+    {
+        const idx = dice(rnd_gen, stakes.map!(stake => stake.amount.integral));
+
+        auto qnode = stakes[idx];
+        if (qnode.key == node_key || added_nodes[idx]) // skip self or dupe
+            continue;
+
+        quorum.nodes ~= qnode.key;
+        added_nodes[idx] = true;
+        if (!quorum_sum.add(qnode.amount))
+            assert(0);
+    }
+
+    const majority = max(1, cast(size_t)ceil(0.67 * quorum.nodes.length));
+    quorum.threshold = majority;
+
+    // we want a predictable order of nodes
+    sort(quorum.nodes);
+
+    return quorum;
+}
+
+/*******************************************************************************
+
+    Verify that the provided quorum sets are considered sane by SCP.
+
+    The quorums are checked both pre and post-normalization,
+    with extra safety checks enabled.
+
+    Params:
+        quorums = the quorum map of (node => quorum) to verify
+
+    Throws:
+        AssertError if the quorum is not considered sane by SCP.
+
+*******************************************************************************/
+
+private void verifyQuorumsSanity (QuorumConfig[PublicKey] quorums)
+{
+    import scpd.scp.QuorumSetUtils;
+
+    foreach (key, quorum; quorums)
+    {
+        auto scp_quorum = toSCPQuorumSet(quorum);
+        const(char)* reason;
+
+        assert(isQuorumSetSane(scp_quorum, true, &reason),
+            format("Key %s with %s fails sanity check before normalization: %s",
+                    key.nice, quorum.toToml, reason.to!string));
```

Comment:
This can't be used here. `nice` should only be available in unittests, using it in production we'll quickly run out of key => name maps.